### PR TITLE
[SPARK-28659][SQL] Use data source if convertible in insert overwrite directory

### PR DESCRIPTION
**Pull request by Udbhav30**
### What changes were proposed in this pull request?
In insert overwrite directory while using `STORED AS file_format`,  files are not compressed.
In this PR it is converted to `datasource`  if it is convertible, to make it inline with`CTAS` behavior which is fixed in this  [PR](https://github.com/apache/spark/pull/22514)

### Why are the changes needed?
To make the behavior inline with `CTAS` while using `STORED AS file_format`

### Does this PR introduce any user-facing change?
Yes, After the fix of this PR now `STORED AS file_format` will be converted to `datasource` if it is convertible 
**Before**
before

**After**
after

### How was this patch tested?
New testcase is added

### CODE REVIEW

**Comment by maropu** [comment]
Can you add tests?

**Comment by Udbhav30** [comment]
> Can you add tests?

yes i will add tests and update the PR

**Comment by Udbhav30** [comment]
@maropu added the testcase!

**Comment by maropu** [comment]
ok to test

**Comment by maropu** [comment]
Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 1]
```diff
@@ -1396,6 +1396,17 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder(conf) {
       compressed = false,
       properties = rowStorage.properties ++ fileStorage.properties)
 
-    (ctx.LOCAL != null, storage, Some(DDLUtils.HIVE_PROVIDER))
+    val fileFormat = extractFileFormat(fileStorage.serde)
+    (ctx.LOCAL != null, storage, Some(fileFormat))
+  }
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]:
```
    val fileFormat = fileStorage.serde.getOrElse(DDLUtils.HIVE_PROVIDER)
    (ctx.LOCAL != null, storage, fileFormat)
```
?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 2]
```diff
@@ -441,7 +441,6 @@ class InsertSuite extends DataSourceTest with SharedSQLContext {
         sql(
           s"""
              |INSERT OVERWRITE LOCAL DIRECTORY '$path'
-             |STORED AS orc
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala]: Why did you delete this?

Comment (Udbhav30) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala]: if we use `stored as` and file format is orc or parquet it will be converted to data source flow.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 1 above)
Comment (Udbhav30) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: in case of `parquet` and `orc` we can use the respective file format instead of `hive`.
In case of ctas also we convert to use data source https://github.com/viirya/spark-1/blob/839a6ce1732fa37b5f8ec9afa2d51730fc6ca691/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala#L188  
This will make it inline with that behavior.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala (hunk 2 above)
Comment (advancedxy) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala]: I believe the test case name should be modified correspondingly and a new test case for (orc/parquet) should be added.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 3]
```diff
@@ -1396,6 +1396,17 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder(conf) {
...
+
+  private def extractFileFormat(serde: Option[String]): String = {
+    if (serde.toString.contains("parquet")) {
```
Comment (advancedxy) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Although `None.toString` is `"None"`, this looks like a hack to me.

How about:
```
serde.map({ x =>
  val lowerCaseSerde = x.toLowerCase(Locale.ROOT)
  if (lowerCaseSerde.contains("parquet") "parquet"
  else if (lowerCaseSerde.contains("orc") "orc"
  else  DDLUtils.HIVE_PROVIDER
}).getOrElse( DDLUtils.HIVE_PROVIDER)
```

**Review by advancedxy — CHANGES_REQUESTED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala (hunk 2 above)
Comment (Udbhav30) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala]: @advancedxy i have updated the testcase name and for orc and parquet testcase was already added. As they will be converted to data source,  the data in the directory would be compressed.

**Comment by Udbhav30** [comment]
@maropu  @advancedxy Gentle ping

**Comment by Udbhav30** [comment]
cc @HyukjinKwon

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by Udbhav30** [comment]
failed test doesn't look related to this PR @HyukjinKwon

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by Udbhav30** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 4]
```diff
@@ -1396,6 +1396,16 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder(conf) {
       compressed = false,
       properties = rowStorage.properties ++ fileStorage.properties)
 
-    (ctx.LOCAL != null, storage, Some(DDLUtils.HIVE_PROVIDER))
+    val fileFormat = extractFileFormat(fileStorage.serde)
+    (ctx.LOCAL != null, storage, Some(fileFormat))
+  }
+
+  private def extractFileFormat(serde: Option[String]): String = {
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: I think @maropu has kind of its point. Why don't you just use `STORED AS file_format` instead? BTW, I think it should be in `RelationConversions` with `spark.sql.hive.convertMetastoreXXX`.

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Yea, I think so... the third param is mainly used for data source tables. In this hive path, this param is just used for a flag, a hive table or not.

Comment (Udbhav30) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: @HyukjinKwon I used `STORED AS file_format` only.. if we use `STORED AS` it will be hive managed, so in this PR i  convert it to data source if convertible, we follow same behavior  in case of `CTAS` which is fixed in ` SPARK-25271`

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: @Udbhav30, I more meant this seems not the right place to replace. You should add a configuration, and such replacement should be done in analysis/optimizing, not in the parser.

If that's the case, then why don't you use `USING file_format` explicitly? Can you please describe what this PR target to fix clearly?

**Comment by HyukjinKwon** [comment]
If you only target to fix Hive ser/de to respect configuration, why don't you set Hive compression properly?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 4 above)
Comment (Udbhav30) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: @HyukjinKwon, i can use `USING file_format` explicitly that would serve the purpose but i thought it is better to fix this and make it inline with `CTAS` behavior  which is fixed in this [PR](https://github.com/apache/spark/pull/22514)

If you agree to go ahead i can try making changes in analysis/optimizing layer instead of parser as suggested by you. 
Sure i will follow the template and update the PR description. thanks :)

**Comment by Udbhav30** [comment]
> If you only target to fix Hive ser/de to respect compression, why don't you set Hive compression properly?
Yes compression can be achieved by setting Hive ser/de or `USING file_format', but as i mentioned this PR is more towards making the behavior inline to `CTAS` and co use `datasource` if it is convertible. Let me know if you have any suggestions :)

**Review by HyukjinKwon — COMMENTED** [review]
Alright, if we want to do the conversion, it should have a configuration like `spark.sql.hive.convertMetastoreCtas` which respects `spark.sql.hive.convertMetastoreParquet` or `spark.sql.hive.convertMetastoreOrc`.

cc @viirya, @dongjoon-hyun, @cloud-fan
